src/charm.py

```diff
@@ -19,30 +21,48 @@
 class MysqlOperatorCharm(CharmBase):
     """A Juju Charm for MySQL InnoDB Cluster."""
 
+    _stored = StoredState()
+
     def __init__(self, *args):
         super().__init__(*args)
 
-        self.framework.observe(self.on.start, self._on_start)
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
-        change_events = [
+        self._src_dir = Path(__file__).parent
+        self._server_container = "mysql-server"
+        self._router_container = "mysql-router"
+        self._stored.set_default(root_password="", mysql_password="")
+
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.leader_elected, self._on_leader_elected)
+        pebble_events = [
             self.on.mysql_server_pebble_ready,
             self.on.mysql_router_pebble_ready,
         ]
-        for event in change_events:
+        for event in pebble_events:
             self.framework.observe(event, self._on_pebble_ready)
 
     # ---- Event handlers
-    def _on_start(self, _):
-        """Placeholder for StartEvent."""
-        pass
+    def _on_install(self, event: WorkloadEvent) -> None:
+        """Ensures the MySQL server instance is configured for an InnoDB Cluster."""
+        container = self.unit.get_container(self._server_container)
+        if not container.can_connect():
+            event.defer()
+            return
 
-    def _on_config_changed(self, _):
-        """Placeholder for ConfigChangedEvent."""
-        pass
+        # Configure server instance:
+        # Push configuration file and instance script to workload container
+        self._mysql.configure_server_instance()
+
+    def _on_leader_elected(self, event: WorkloadEvent) -> None:
+        """Set all needed passwords for connection and administration."""
+        self._stored.root_password = self._root_password
+        self._stored.mysql_password = self._mysql_password
```

Comment:
Since this information lives in the config, you should not track it in the StoredState. 

The Operator Framework loads an up-to-date copy of the config when it starts, at the beginning of each invocation. You can rely on that data being accurate, and low cost to access.